File: search-suggestions-system/search-suggestions-system.java
class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        List<List<String>> result = new ArrayList<>();
        
        Arrays.sort(products); 
        
        int n = products.length;
        int low = 0;  
        int high = n - 1; 
        int len = searchWord.length();  
        
        for (int i = 0; i < len; i++){   
            while (low <= high && (products[low].length() <= i || products[low].charAt(i) != searchWord.charAt(i))){
                low++;
            }                      
            while (low <= high && (products[high].length() <= i || products[high].charAt(i) != searchWord.charAt(i))){
                high--;
            }
            
            int min = Math.min(low + 3, high + 1);
            List<String> list = new ArrayList<>();
            for (int j = low; j < min; j++){
                list.add(products[j]);
            }
            result.add(list);    
        }
        return result;
    }
}

File: 1397-search-suggestions-system/search-suggestions-system.py
class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        products.sort()
        result = []

        for i in range(1, len(searchWord) + 1):
            prefix = searchWord[:i]

            idx = bisect_left(products, prefix)
            suggestions = []

            for j in range(idx, min(idx + 3, len(products))):
                if products[j].startswith(prefix):
                    suggestions.append(products[j])
            result.append(suggestions)
        return result